### PR TITLE
Manager for `cco ckan env` and logic for it

### PR DESCRIPTION
**Pull request by zelima**

### CODE REVIEW

File: ckan_cloud_operator/providers/ckan/env/manager.py [hunk 1]
```diff
@@ -22,373 +21,159 @@
...
+        for _env in envs:
+            _env['active'] = False
+        kwargs['active'] = True
+        kwargs['name'] = environment
+        envs.append(kwargs)
+        cco_config['environments'] = envs
+        # update config file
+        _mkconfdir()
+        _write_yaml(cco_config)
+        print(f'{kwargs["cluster_name"]} has succefully configred for "{environment}" environment')
```
Comment (pdelboca) [inline ckan_cloud_operator/providers/ckan/env/manager.py]:
```suggestion
        print(f'{kwargs["cluster_name"]} has been successfully configured for "{environment}" environment')
```

File: ckan_cloud_operator/providers/ckan/env/manager.py [hunk 2]
```diff
@@ -22,373 +21,159 @@
...
-    if instance_name:
-        set_name(instance_id, instance_name, dry_run=dry_run)
+def update_environment(environment, **kwargs):
+    updated_envs = []
+    cco_config = {}
+    if _cco_is_configured() and _get_config():
+        cco_config = _get_config()
+        env_exists = any([environment.lower() == _env.get('name', '').lower() for _env in cco_config.get('environments') or []])
+        if not env_exists:
+            logs.info(f'Environment {environment} does not exists please run `cco ckan env add` with the same flags')
```
Comment (pdelboca) [inline ckan_cloud_operator/providers/ckan/env/manager.py]:
```suggestion
            logs.info(f'Environment {environment} does not exists. Please run `cco ckan env add` with the same flags')
```

File: ckan_cloud_operator/providers/ckan/env/manager.py [hunk 3]
```diff
@@ -22,373 +21,159 @@
...
 
-    return instance_id
+def set_environment(environment):
+    updated_envs = []
+    cco_config = {}
+    if _cco_is_configured() and _get_config():
+        cco_config = _get_config()
+        env_exists = any([environment.lower() == _env.get('name', '').lower() for _env in cco_config.get('environments') or []])
+        if not env_exists:
+            logs.info(f'Environment {environment} does not exists please run `cco ckan env add` with the same flags')
```
Comment (pdelboca) [inline ckan_cloud_operator/providers/ckan/env/manager.py]:
```suggestion
            logs.info(f'Environment {environment} does not exists. Please run `cco ckan env add` with the same flags')
```

File: ckan_cloud_operator/providers/ckan/env/manager.py [hunk 4]
```diff
@@ -22,373 +21,159 @@
...
-        kubectl.apply(resource)
 
+def remove_environment(environment):
+    updated_envs = []
+    cco_config = {}
+    if _cco_is_configured() and _get_config():
+        cco_config = _get_config()
+        env_exists = any([environment.lower() == _env.get('name', '').lower() for _env in cco_config.get('environments') or []])
+        if not env_exists:
+            logs.info(f'Environment {environment} does not exists please run `cco ckan env add {environment}`')
```
Comment (pdelboca) [inline ckan_cloud_operator/providers/ckan/env/manager.py]:
```suggestion
            logs.info(f'Environment {environment} does not exists. Please run `cco ckan env add {environment}`')
```

File: ckan_cloud_operator/providers/ckan/env/manager.py [hunk 5]
```diff
@@ -22,373 +21,159 @@
...
+            return
+        envs = cco_config.get('environments', [])
 
-def set_storage(instance_id, instance_name, dry_run=False):
-    from ckan_cloud_operator.providers.storage.manager import get_provider, get_provider_id
+        for _env in envs:
+            if _env.get('name').lower() == environment.lower():
+                continue
```
Comment (pdelboca) [inline ckan_cloud_operator/providers/ckan/env/manager.py:136]: Is this continue intentional? Or it should be removing environments?

File: ckan_cloud_operator/providers/ckan/env/manager.py [hunk 6]
```diff
@@ -22,373 +21,159 @@
...
+
+def _mkconfdir():
+    dir = _get_config_dir()
+    Path(f'{dir}').mkdir(parents=True, exist_ok=True)
+
+
+def _run_az_command_or_login(command, func, *args):
+    if not args[-1]:
+        logs.subprocess_check_output(['az', 'login'])
+        print("Logged in succefully")
```
Comment (pdelboca) [inline ckan_cloud_operator/providers/ckan/env/manager.py]:
```suggestion
        print("Logged in successfully.")
```

File: ckan_cloud_operator/providers/ckan/env/manager.py [hunk 7]
```diff
@@ -22,373 +21,159 @@
...
+    try:
+        logs.subprocess_check_call(command.split(' '))
+    except AttributeError as e:
+        print('You need to login in to azure CLI. Please authenticate with browser')
```
Comment (pdelboca) [inline ckan_cloud_operator/providers/ckan/env/manager.py]:
```suggestion
        print('You need to login to azure CLI. Please authenticate with browser.')
```

File: ckan_cloud_operator/providers/ckan/env/manager.py [hunk 8]
```diff
@@ -22,373 +21,159 @@
...
+
+
+def list_environments():
+    if _cco_is_configured() and _get_config():
+        cco_config = _get_config()
+        for environment in cco_config.get('environments', []):
+            prefix = '>' if environment.get('active', False) else '-'
+            print(prefix, environment.get('name'))
+    else:
+        print('Ckan Cloud Operator is no yet configred. Please run cco add << environment >>')
```
Comment (pdelboca) [inline ckan_cloud_operator/providers/ckan/env/manager.py]:
```suggestion
        print('Ckan Cloud Operator is no yet configured. Please run cco add << environment >>')
```

**Review by pdelboca — DISMISSED** [review]
Looks good @zelima ! Just a few small typos.

File: ckan_cloud_operator/providers/ckan/env/manager.py (hunk 5 above)
Comment (zelima) [reply ckan_cloud_operator/providers/ckan/env/manager.py:136]: Intentional - this is for removing environments. It collects all but selected

**Pull request merged**